File: 2024/csharp/Puzzles/Dec01.cs
using csharp.utils;

namespace csharp.Puzzles;

public static class Dec01
{
    public static int solvePart1(string? date, bool useTestData = false)
    {
        Console.WriteLine("Day 01, Part 1: Id lists total distance");

        // create a data file reader and read the file.
        var dfr = new DataFileReader(date: date, useTestData: useTestData, part: 1);
        dfr.ReadFile();

        // print the data
        // Console.WriteLine("Test Data:");
        // for (int i = 0; i < dfr.Lines.Count; i++) 
        // {
        //     Console.WriteLine(dfr.Lines[i]);
        // }
        // Console.WriteLine("----------");

        // create two lists of Ids
        List<int> firstIds = new List<int>();
        List<int> secondIds = new List<int>();

        CreateLists(lines: dfr.Lines, first: firstIds, second: secondIds);

        // for each line in the file
        // for (int i = 0; i < dfr.Lines.Count; i++)
        // {
        //     // add the first id to one list and the second id to another list
        //     string[] lineIds = dfr.Lines[i].Split(" ", StringSplitOptions.RemoveEmptyEntries);
            
        //     // Console.WriteLine($"lineIds: {lineIds[0]} {lineIds[1]}");

        //     int id1 = Convert.ToInt32(lineIds[0]);
        //     int id2 = Convert.ToInt32(lineIds[1]);

        //     // Console.WriteLine($"id1: {id1}, id2: {id2}");

        //     firstIds.Add(id1);
        //     secondIds.Add(id2);
        // }

        // sort both lists
        firstIds.Sort();
        secondIds.Sort();

        // compare both lists and count
        if (firstIds.Count != secondIds.Count)
        {
            throw new Exception("Error: firstIds and secondIds are not the same length");
        }

        List<int> diffs = new List<int>();

        for (int i = 0; i < firstIds.Count; i++)
        {
            // take absolute difference
            diffs.Add(firstIds[i] >= secondIds[i] ? firstIds[i] - secondIds[i] : secondIds[i] - firstIds[i]);
        }

        // count the total
        int total = 0;

        for (int i = 0; i < diffs.Count; i++)
        {
            total += diffs[i];
        }
        
        var outputString = useTestData ? "Total [using test data]" : "Total [using puzzle data]";
        Console.WriteLine($"{outputString}: {total}\n\n");

        return total;
    }

    public static int solvePart2(string date, bool useTestData = false)
    {
        Console.WriteLine("Day 01, Part 2: Id lists similarity score");

        // create a data file reader and read the file.
        var dfr = new DataFileReader(date: date, useTestData: useTestData, part: 2);
        dfr.ReadFile();

        // create two lists of Ids
        List<int> firstIds = new List<int>();
        List<int> secondIds = new List<int>();

        CreateLists(lines: dfr.Lines, first: firstIds, second: secondIds);

        // for each line in the file
        // for (int i = 0; i < dfr.Lines.Count; i++)
        // {
        //     // add the first id to one list and the second id to another list
        //     string[] lineIds = dfr.Lines[i].Split(" ", StringSplitOptions.RemoveEmptyEntries);
            
        //     // Console.WriteLine($"lineIds: {lineIds[0]} {lineIds[1]}");

        //     int id1 = Convert.ToInt32(lineIds[0]);
        //     int id2 = Convert.ToInt32(lineIds[1]);

        //     // Console.WriteLine($"id1: {id1}, id2: {id2}");

        //     firstIds.Add(id1);
        //     secondIds.Add(id2);
        // }

        // compare both lists and count
        if (firstIds.Count != secondIds.Count)
        {
            throw new Exception("Error: firstIds and secondIds are not the same length");
        }

        // find the similarity scores for each id in the first list

        List<int> scores = new List<int>();


        for (int i = 0; i < firstIds.Count; i++)
        {
            var count = 0;

            for (int j = 0; j < secondIds.Count; j++)
            {
                if (firstIds[i] == secondIds[j])
                {
                    count++;
                }
            }
            scores.Add(firstIds[i] * count);
        }

        // count the total
        int total = 0;

        for (int i = 0; i < scores.Count; i++)
        {
            total += scores[i];
        }

        var outputString = useTestData ? "Total [using test data]" : "Total [using puzzle data]";
        Console.WriteLine($"{outputString}: {total}\n\n");

        return total;

    }


    private static void CreateLists(List<string> lines, List<int> first, List<int> second)
    {
        // for each line in the file
        for (int i = 0; i < lines.Count; i++)
        {
            // add the first id to one list and the second id to another list
            string[] lineIds = lines[i].Split(" ", StringSplitOptions.RemoveEmptyEntries);
            
            // Console.WriteLine($"lineIds: {lineIds[0]} {lineIds[1]}");

            int id1 = Convert.ToInt32(lineIds[0]);
            int id2 = Convert.ToInt32(lineIds[1]);

            // Console.WriteLine($"id1: {id1}, id2: {id2}");

            first.Add(id1);
            second.Add(id2);
        }

    }







}




        // Part 1: pseudo code
        // for each line in the file

        // add the first id to one list and the second id to another list

        // sort both lists

        // compare both lists and count

        // count the total

File: 2024/ts/src/puzzles/dec01.ts
import { PuzzleSolver } from './../types/index';
import { dataFileReader } from '../utils/index';


const solvePart1: PuzzleSolver = (date, useTestData = false) => {
	console.log('Day 01, Part 1');

	const lines = dataFileReader({ date, useTestData, part: 1 });

	// console.log('lines: ');
	// lines.map(line => console.log(line));
	// console.log('--------------------\n');

	// create and sort lists
	const firstIds = lines.map(line => parseInt(line.split(' ')[0])).sort();
	const secondIds = lines.map(line => parseInt(line.split(' ').filter(part => part !== '')[1])).sort();

	// compare both lists and count
	if (firstIds.length !== secondIds.length) {
		console.log('Error: lists are not the same length');
		return;
	}

	const diffs: number[]  = [];

	for (let i = 0; i < firstIds.length; i++) {
		diffs.push(firstIds[i] >= secondIds[i] ? (firstIds[i] - secondIds[i]) : (secondIds[i] - firstIds[i]));
	}

	const total = diffs.reduce((acc, curr) => acc + curr, 0);

	var outputString = useTestData ? "Total [using test data]" : "Total [using puzzle data]";
	console.log(`${outputString}: ${total}\n\n`);

	
};

const solvePart2: PuzzleSolver = (date, useTestData = false) => {
	console.log('Day 01, Part 2');

	const lines = dataFileReader({ date, useTestData, part: 1 });

		// create lists
		const firstIds = lines.map(line => parseInt(line.split(' ')[0]));
		const secondIds = lines.map(line => parseInt(line.split(' ').filter(part => part !== '')[1]));

		// compare both lists and count
		if (firstIds.length !== secondIds.length) {
			console.log('Error: lists are not the same length');
			return;
		}

		const scores: number[] = [];

		for (let i = 0; i < firstIds.length; i++) {
			let count = 0;
			for (let j = 0; j < secondIds.length; j++) {
				if (firstIds[i] === secondIds[j]) {
					count++;
				}
			}
			scores.push(firstIds[i] * count);
		}

		const total = scores.reduce((acc, curr) => acc + curr, 0);

		var outputString = useTestData ? "Total [using test data]" : "Total [using puzzle data]";
		console.log(`${outputString}: ${total}\n\n`);

};

export default {
	solvePart1,
	solvePart2
}